### PR TITLE
QuestionBot Expansion

### PR DESCRIPTION
**Pull request by itdaniher**
Adds `do_challenge`, which serves as a basic captcha, and `ask_intable` / `ask_floatable` abstractions which prompt the user if an improperly formatted value is provided.

Also adds `require_first_device` flag to all `ask_` implementations, which will tell user to message from their phone if they aren't.

### CODE REVIEW

**Comment by itdaniher** [comment]
The non-fallback captcha is separately PRed in #124 .

File: forest/core.py [hunk 1]
```diff
@@ -1072,8 +1077,17 @@ class QuestionBot(PayBot):
...
+        self.requires_first_device: dict[str, bool] = {}
+        self.terminal_answers = "stop quit exit break cancel".split()
+        self.failed_user_challenges: dict[str, int] = {}
+        self.FIRST_DEVICE_PLEASE = "Please answer from your phone or primary device!"
         super().__init__(bot_number)
 
+    def is_first_device(self, msg: Message) -> bool:
+        if not msg.blob:
+            return False
+        return msg.blob.get("remote_address", {}).get("device_id", 0) == 1
```
Comment (iamalwaysuncomfortable) [inline forest/core.py:1089]: This is great

File: forest/core.py [hunk 2]
```diff
@@ -1090,51 +1112,151 @@ async def handle_message(self, message: Message) -> Response:
...
     ) -> str:
         await self.send_message(recipient, question_text)
         answer_future = self.pending_answers[recipient] = asyncio.Future()
+        if require_first_device:
+            self.requires_first_device[recipient] = True
         answer = await answer_future
         self.pending_answers.pop(recipient)
         return answer.full_text or ""
 
+    async def ask_floatable_question(
```
Comment (iamalwaysuncomfortable) [inline forest/core.py:1165]: Perhaps condense intable and floatable into ask_numeric_question and have int or float specified in a required method parameter?

**Review by iamalwaysuncomfortable — COMMENTED** [review]
Cool abstractions. Perhaps in an inhereted class, dialogs can be specified as lists of questions with answers which when triggered are iterated through?

Also what are thoughts on state abstractions. If you trigger a dialog and set things, how do we track? Questions for soon.

File: forest/core.py (hunk 2 above)
Comment (itdaniher) [reply forest/core.py:1165]: different behaviors alas, in that floatable needs to accept ',' or '.' and intable only accepts integers.

File: forest/core.py (hunk 1 above)
Comment (itdaniher) [reply forest/core.py:1089]: thanks!!

File: forest/core.py [hunk 3]
```diff
@@ -1082,6 +1096,14 @@ async def handle_message(self, message: Message) -> Response:
             probably_future = self.pending_answers.get(
                 message.uuid
             ) or self.pending_answers.get(message.source)
+            if (
+                message.uuid in self.requires_first_device
+                or message.source in self.requires_first_device
```
Comment (technillogue) [inline forest/core.py:1101]: any chance we could come up with a cleaner pattern for this? it shows up in a lot of places `(message.uuid or message.source) in dict` could do it, and might be nicer to use going forward if nothing uses number as a key when uuid is available. might be a more involved solution featuring a class for uuid/source that implements __hash__ and/or __eq__

File: forest/core.py (hunk 2 above)
Comment (technillogue) [reply forest/core.py:1165]: this is okay but i could see combining ask_{floatable,intable}_question them with the approach Mikey used with, with shorter functions handle ','/'.' logic. that has the downside of maybe making it harder to follow up with "(as a decimal, ie 1.01 or 2,02)" and  "(as a whole number, ie '1' or '2')"

File: forest/core.py [hunk 4]
```diff
@@ -1090,51 +1112,151 @@ async def handle_message(self, message: Message) -> Response:
...
+        if answer.full_text and not answer.full_text.isnumeric():
+            if answer.full_text.lower() in self.terminal_answers:
+                return None
+            if question_text and "as a whole number" in question_text:
+                return await self.ask_intable_question(recipient, question_text)
+            return await self.ask_intable_question(
+                recipient, (question_text or "") + " (as a whole number, ie '1' or '2')"
+            )
+        if answer.full_text:
+            return int(answer.full_text)
```
Comment (technillogue) [inline forest/core.py:1216]: no point in checking for "1.00" ig, but "1,000,000" might be relevant? worse still some SI places might consider "1.000.000" to be the same number.

Comment (technillogue) [reply forest/core.py:1216]: not sure if `try: int(...) except: return await self.ask..` might be cleaner

Comment (technillogue) [reply forest/core.py:1216]: also "$1.00" could be an int, but we probably really don't need to catch that one

File: forest/core.py [hunk 5]
```diff
@@ -1090,51 +1112,151 @@ async def handle_message(self, message: Message) -> Response:
...
+        # the captcha module delivers graphical challenges of the same format
+        if captcha is not None:
+            challenge, answer = captcha.get_challenge_and_answer()
+            await self.send_message(
+                msg.uuid,
+                "Please answer this arithmetic problem to prove you're (probably) not a bot!",
+                attachments=[challenge],
+            )
+        else:
+            # I really don't like Python's random module, but this could be cross-platform some other way
```
Comment (technillogue) [inline forest/core.py:1244]: https://docs.python.org/3/library/secrets.html will apparently use  CryptGenRandom() on windows and /dev/urandom otherwise

**Review by technillogue — APPROVED** [review]
some nitpicks but i would be okay with this going in

File: forest/core.py (hunk 3 above)
Comment (itdaniher) [reply forest/core.py:1101]: Users will sporadically use `message.source` / phone numbers in one way or another, either by typing manually or error. Might be able to abstract over.

File: forest/core.py (hunk 5 above)
Comment (itdaniher) [reply forest/core.py:1244]: Thanks!

File: forest/core.py (hunk 4 above)
Comment (itdaniher) [reply forest/core.py:1216]: added an example showing 2000 written without points or commas for reference.

File: forest/core.py (hunk 3 above)
Comment (itdaniher) [reply forest/core.py]: Refactored into a helper function. Longer term we'd benefit from a class for User and a class for Question that includes the first device requirement and optionally abstracts over CAPTCHA challenges, etc.

**Pull request merged**